docs(ai-proxy-advanced): add new feature description for ai-proxy-advanced in 3.10 (#8583)
Kong/docs.konghq.com: app/_hub/kong-inc/ai-proxy-advanced/overview/_index.md

* **Lowest-usage**: The lowest-usage algorithm in AI Proxy Advanced is based on the volume of usage for each model. It balances the load by distributing requests to models with the lowest usage, measured by factors such as prompt token counts, response token counts, {% if_version gte:3.10.x %} cost, {% endif_version %} or other resource metrics.
* **Lowest-latency**: The lowest-latency algorithm is based on the response time for each model. It distributes requests to models with the lowest response time.
* **Semantic**: The semantic algorithm distributes requests to different models based on the similarity between the prompt in the request and the description provided in the model configuration. This allows Kong to automatically select the model that is best suited for the given domain or use case. This feature enhances the flexibility and efficiency of model selection, especially when dealing with a diverse range of AI providers and models.
* [Round-robin (weighted)](/gateway/latest/how-kong-works/load-balancing/#round-robin)
* [Consistent-hashing (sticky-session on given header value)](/gateway/latest/how-kong-works/load-balancing/#consistent-hashing)

> {% if_version gte:3.10.x %}

Reviewer: ```suggestion
{% if_version gte:3.10.x -%}
```

* [Priority Group](/gateway/latest/how-kong-works/load-balancing/#priority)
{% endif_version %}

## Retry and fallback

The load balancer has customizable retries and timeouts for requests, and can redirect a request to a different model in case of failure. This allows you to have a fallback in case one of your targets is unavailable.

{% if_version gte:3.10.x %}
This plugin supports fallback over targets with any supported formats.
{% else %}
This plugin does not support fallback over targets with different formats. You can use different providers as long as the formats are compatible. For example, load balancers with these combinations of targets are supported:
* Different OpenAI models
* OpenAI models and Mistral models with the OpenAI format
* Mistral models with the OLLAMA format and Llama models with the OLLAMA format
{% endif_version %}

{:.note}
{% if_version gte:3.10.x %}

> To configure failover in addition to network errors, set `config.balancer.failover_criteria` to include additional HTTP error codes, like `"http_429"` or `"http_502"`. Note that `"non_idenpotent"` should be set as well as most AI services accepts

Reviewer 2: `non_idenpotent` is for the HTTP POST method of these LLM APIs, Do we need to describe these details here? `should be set as well as most AI services accepts` seems ambiguous.

Author: Yes, I was trying to describe "because almost all LLM APIs only accepts POST requests so we always need to set `non_denpotent`". I wil leave it for docs team to help me polish this, I'm not good at describing this

Reviewer: https://github.com/Kong/docs.konghq.com/pull/8583#discussion_r2001449975 fixed it here

> To configure failover in addition to network errors, set `config.balancer.failover_criteria` to include additional HTTP error codes, like `"http_429"` or `"http_502"`. Note that `"non_idenpotent"` should be set as well as most AI services accepts
> POST requests.

Reviewer: ```suggestion
> To configure failover in addition to network errors, set [`config.balancer.failover_criteria`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-balancer-failover_criteria) to include:
* Additional HTTP error codes, like `http_429` or `http_502`
* The `non_idempotent` setting, as most AI services accept POST requests
```

> While only the **Llama2** and **Mistral** models are classed as self-hosted, the target URL can be overridden for any of the supported providers.
> For example, a self-hosted or otherwise OpenAI-compatible endpoint can be called by setting the same [`config.model.options.upstream_url`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-embeddings-model-options-upstream_url) plugin option.

{% if_version gte:3.10.x %}

> Furthermore, if you are using each provider's native SDK, Kong allows to transparently proxy pass the

Reviewer: ```suggestion
If you are using each provider's native SDK, {{site.base_gateway}} allows you to transparently proxy the
```

> request without any transformation and return the response unmodified. This can be done by setting
> [`config.llm_format`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-llm_format) to a value other than `"openai"`, such as `"gemini"` or `"bedrock"`.

Reviewer: ```suggestion
request without any transformation and return the response unmodified. This can be done by setting
[`config.llm_format`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-llm_format) to a value other than `openai`, such as `gemini` or `bedrock`.
```

> In this mode, Kong will still provider useful analytics, logging and cost calculation.

Reviewer: ```suggestion
In this mode, {{site.base_gateway}} will still provide useful analytics, logging, and cost calculation.
```

{% endif_version %}

### Input formats

> Kong will mediate the request and response format based on the selected [`config.targets[].provider`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-embeddings-model-provider) and [`config.targets[].route_type`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-targets-route_type), as outlined in the table above.

Reviewer 2: This change looks like it also needs to be judgment by kong version
```
{% if_version gte:3.10.x %}
```

Author: This is okay, it's a copy-paste typo from ai-proxy. AI Proxy Advanced plugin uses targets[] array from the first version.

> Kong will mediate the request and response format based on the selected [`config.targets[].provider`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-embeddings-model-provider) and [`config.targets[].route_type`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-targets-route_type), as outlined in the table above.
> 
> {% if_version gte:3.10.x %}
> By default, Kong uses OpenAI format and can be changed using [`config.targets[].llm_format`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-targets-llm_format). When `llm_format` is not set to `openai`, this plugin will not transform the request when sending to upstream and keep as is.
> {% endif_version %}

Reviewer: ```suggestion
{% if_version lte:3.9.x %}
Kong will mediate the request and response format based on the selected [`config.provider`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-embeddings-model-provider) and [`config.route_type`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-targets-route_type), as outlined in the table above.
{% endif_version %}

{% if_version gte:3.10.x %}
Kong will mediate the request and response format based on the selected [`config.targets[].provider`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-embeddings-model-provider) and [`config.targets[].route_type`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-targets-route_type), as outlined in the table above.

By default, Kong uses OpenAI format and can be changed using [`config.targets[].llm_format`](/hub/kong-inc/ai-proxy-advanced/configuration/#config-targets-llm_format). When `llm_format` is not set to `openai`, this plugin will not transform the request when sending to upstream and will keep it as-is.
{% endif_version %}
```
